### PR TITLE
Support non correlated scalar subqueries.

### LINKED ISSUE #2878: Add support for scalar subqueries
Scalar sub queries are not supported.  For example:
```sql
SELECT *
FROM nation
WHERE nation KEY >= (SELECT nation KEY FROM nation WHERE NAME ='ALGERIA' LIMIT 1)
```

### PR DESCRIPTION
**Pull request by kokosing**
Support non correlated scalar subqueries.

Scalar subqueries are rewritten into a join in a following manner.
From:
  `SELECT * FROM nation WHERE regionkey = (
      SELECT regionkey FROM region WHERE name = 'AFRICA'
   )`
into:
  `SELECT * FROM nation, (
      SELECT regionkey FROM region WHERE name = 'AFRICA'
   ) temp
  WHERE nation.regionkey = temp.regionkey`

Also above subquery plan is wrapped in a ScalarNode which goal is to
maintain scalar subquery returns at most single value. To achieve that
during plan fragmentation, there is a plan fragment generated
from ScalarNode so all the results of the subquery goes through one
node.

Fixes #2878.

### CODE REVIEW

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 1]
```diff
@@ -402,14 +402,14 @@ public void testUnnest()
         assertQuery("SELECT count(*) FROM UNNEST(ARRAY[1, 2, 3], ARRAY[4, 5])", "SELECT 3");
         assertQuery("SELECT a FROM UNNEST(ARRAY['kittens', 'puppies']) t(a)", "SELECT * FROM VALUES ('kittens'), ('puppies')");
         assertQuery("" +
-                "SELECT c " +
-                "FROM UNNEST(ARRAY[1, 2, 3], ARRAY[4, 5]) t(a, b) " +
-                "CROSS JOIN (values (8), (9)) t2(c)",
+                        "SELECT c " +
```
Comment (losipiuk) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: revert whitespace

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 2]
```diff
@@ -4391,6 +4392,20 @@ public void testSemiJoinNullHandling()
     }
 
     @Test
+    public void testScalarSubquery()
+            throws Exception
+    {
+        assertQuery("SELECT (SELECT (SELECT (SELECT 1)))");
+        assertQuery("select * from lineitem where orderkey = (\n" +
```
Comment (losipiuk) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: use uppercase SQL keywords

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java [hunk 3]
```diff
@@ -182,7 +197,7 @@ protected void assertAccessDenied(
             fail("Expected " + AccessDeniedException.class.getSimpleName());
         }
         catch (RuntimeException e) {
-            assertExceptionMessage(e, exceptionsMessageRegExp);
+            assertExceptionMessage(e, ".*Access Denied: " + exceptionsMessageRegExp);
```
Comment (losipiuk) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java]: unrelated?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 4]
```diff
@@ -307,10 +307,9 @@ public FragmentProperties setPartitionedOutput(Optional<List<Symbol>> partitionK
 
             if (partitionKeys.isPresent()) {
                 this.outputPartitioning = Optional.of(OutputPartitioning.HASH);
-            this.nullPartitionPolicy = Optional.of(NullPartitioning.HASH);
+                this.nullPartitionPolicy = Optional.of(NullPartitioning.HASH);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: extract to fix whitespace commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 5]
```diff
@@ -105,8 +108,8 @@
     protected PlanBuilder visitQuery(Query query, Void context)
     {
         PlanBuilder builder = planQueryBody(query);
-        Set<InPredicate> inPredicates = analysis.getInPredicates(query);
-        builder = appendSemiJoins(builder, inPredicates);
+        builder = appendSemiJoins(builder, analysis.getInPredicates(query));
+        builder = appendScalarSubqueryJoins(builder, analysis.getScalarSubqueries(query));
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: filter out subqueries from IN.
Add test which catches this issue.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 6]
```diff
@@ -145,6 +148,17 @@ protected PlanBuilder visitQuerySpecification(QuerySpecification node, Void cont
         return builder;
     }
 
+    private Set<SubqueryExpression> getSubqueryExpressionsNotUsedInInPredicates(QuerySpecification node)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: maybe move this method to `Analysis`. And replace the method `getScalarSubqueries`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 7]
```diff
@@ -145,6 +148,17 @@ protected PlanBuilder visitQuerySpecification(QuerySpecification node, Void cont
...
+    {
+        Set<Expression> inPredicateValueLists = analysis.getInPredicates(node).stream()
+                .map(inPredicate -> inPredicate.getValueList())
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: maybe add commit which renames getValueList to something.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 8]
```diff
@@ -819,6 +797,7 @@ protected Type visitSubqueryExpression(SubqueryExpression node, AnalysisContext
             Type type = Iterables.getOnlyElement(descriptor.getVisibleFields()).getType();
 
             expressionTypes.put(node, type);
+            scalarSubqueries.add(node);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: this are all subqueries - not only scalar.
This method throws SemanticException if it spots subquery which returns multiple columns. But that does not ensure that subquery is scalar as it still may return multiple rows.

Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: As we discussed we should change visitSubqueryExpression so it is aware of the position of the subquery being analysed in expression tree.

Then different checks would be done for 
1) subqueries which are right hand side of `IN` operator. We require single column queries for it.
2) subqueries which are used as scalar value (e.g. `x > (SELECT ...). Here we require scalar subquery.

And then different subqueries groups would be stored on different fields in `ExpressionAnalysis` object.

File: presto-main/src/main/java/com/facebook/presto/operator/ScalarOperator.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        try {
+            if (page != null && page.getPositionCount() > 0) {
+                rowsCount += page.getPositionCount();
+                if (rowsCount > 1) {
+                    throw new PrestoException(USER_ERROR, "Scalar sub-query has returned multiple rows");
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/ScalarOperator.java]: This should have a specific error code

Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/ScalarOperator.java]: `SUBQUERY_MULTIPLE_ROWS`

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/ScalarOperator.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: done

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java (hunk 3 above)
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java]: Relates as I use ``assertExceptionMessage`` in ``assertQueryFails`` which I added.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java (hunk 8 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java (hunk 5 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: done

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 2 above)
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: done, this method was removed as it is no longer needed

**Comment by kokosing** [comment]
Rebased, comments addressed, some new tests and fixes.

@electrum @martint @losipiuk Please do a review.

**Comment by kokosing** [comment]
@martint ping

**Comment by kokosing** [comment]
rebased

**Comment by kokosing** [comment]
@martint  ping

**Comment by martint** [comment]
@kokosing, back to you for comments

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 10]
```diff
@@ -4531,6 +4532,23 @@ public void testSemiJoinNullHandling()
     }
 
     @Test
+    public void testScalarSubquery()
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: We're probably going to need more tests.
* multiple scalar subqueries in a single query
* combination of IN predicate and scalar subqueries in the same query
* combinations with other features (joins -- all types, distinct, etc)

**Comment by kokosing** [comment]
@martint I applied all your comments and rebased the patch.

**Comment by martint** [comment]
Merged, thanks!
